lib/util.js

```diff
@@ -282,612 +365,515 @@ inspect.styles = Object.assign(Object.create(null), {
-  if (typeof raw === 'string') {
-    formatted = formatPrimitiveNoColor(ctx, raw);
-    base = ` [String: ${formatted}]`;
-  }
-
-  // Make boxed primitive Numbers look like such
-  if (typeof raw === 'number') {
-    formatted = formatPrimitiveNoColor(ctx, raw);
-    base = ` [Number: ${formatted}]`;
-  }
-
-  // Make boxed primitive Booleans look like such
-  if (typeof raw === 'boolean') {
-    formatted = formatPrimitiveNoColor(ctx, raw);
-    base = ` [Boolean: ${formatted}]`;
-  }
-
-  // Add constructor name if available
-  if (base === '' && constructor)
-    braces[0] = `${constructor.name} ${braces[0]}`;
-
-  if (empty === true) {
-    return `${braces[0]}${base}${braces[1]}`;
-  }
+  // Using an array here is actually better for the average case than using
+  // a Set. `seen` will only check for the depth and will never grow to large.
+  if (ctx.seen.indexOf(value) !== -1)
+    return ctx.stylize('[Circular]', 'special');
 
-  if (recurseTimes < 0) {
-    if (isRegExp(value)) {
-      return ctx.stylize(regExpToString.call(value), 'regexp');
-    } else if (Array.isArray(value)) {
-      return ctx.stylize('[Array]', 'special');
-    } else {
+  if (recurseTimes != null) {
+    if (recurseTimes < 0) {
+      if (Array.isArray(value))
+        return ctx.stylize('[Array]', 'special');
       return ctx.stylize('[Object]', 'special');
     }
-  }
-
-  // TODO(addaleax): Make `seen` a Set to avoid linear-time lookup.
-  if (ctx.seen.includes(value)) {
-    return ctx.stylize('[Circular]', 'special');
+    recurseTimes -= 1;
   }
 
   ctx.seen.push(value);
-  const output = formatter(ctx, value, recurseTimes, visibleKeys, keys);
+  const output = formatter(ctx, value, recurseTimes, keys);
+
+  for (var i = 0; i < symbols.length; i++) {
+    output.push(formatProperty(ctx, value, recurseTimes, symbols[i], 0));
+  }
   ctx.seen.pop();
 
-  return reduceToSingleString(output, base, braces, ctx.breakLength);
+  return reduceToSingleString(ctx, output, base, braces, ln);
 }
 
-
-function formatNumber(ctx, value) {
+function formatNumber(fn, value) {
   // Format -0 as '-0'. Strict equality won't distinguish 0 from -0.
   if (Object.is(value, -0))
-    return ctx.stylize('-0', 'number');
-  return ctx.stylize(`${value}`, 'number');
+    return fn('-0', 'number');
+  return fn(`${value}`, 'number');
 }
 
-
-function formatPrimitive(ctx, value) {
-  if (value === undefined)
-    return ctx.stylize('undefined', 'undefined');
-
-  // For some reason typeof null is "object", so special case here.
-  if (value === null)
-    return ctx.stylize('null', 'null');
-
-  var type = typeof value;
-
-  if (type === 'string') {
-    var simple = JSON.stringify(value)
-                     .replace(/^"|"$/g, '')
-                     .replace(/'/g, "\\'")
-                     .replace(/\\"/g, '"');
-    return ctx.stylize(`'${simple}'`, 'string');
-  }
+function formatPrimitive(fn, value) {
+  const type = typeof value;
```

Comment:
You might double check the performance with this. I know this would cause a performance hit when saving a `typeof` to a variable and checking that variable because V8 used to (and maybe still) would *inline* such checks when doing `typeof foo === 'string'`, but it cannot do that when it is saved to a variable.

Reply:
I checked and my results showed exactly no difference for one way or the other. I have the feeling quite some people prefer always using typeof though, so I went ahead and changed it.